File: 이준영/백준/Gold/1941. 소문난 칠공주/소문난 칠공주.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {
	
	static int count = 0;
	static char[][] arr = new char[5][5];
	static boolean[][] visited = new boolean[5][5];
	static int[] dx = {0,0,1,-1};
	static int[] dy = {1,-1,0,0};
	
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		
		for(int i=0;i<5;i++) {
			String str = sc.nextLine();
			for(int j=0;j<5;j++) {
				arr[i][j]=str.charAt(j);
			}
		}
		btk(0,0,0,-1);
		System.out.println(count);
	}
	
	static void btk(int cur,int c_len,int y_count,int start) {	
		if(y_count>=4) {
			return;
		}
		
		if(c_len==7) {
			if(check(start)) {
				count+=1;
			}
			return;
		}
		if(cur==25) {
			return;
		}
		btk(cur+1,c_len,y_count,start);
		int y = cur/5;
		int x = cur%5;
		if(c_len==0) {
			visited[y][x] = true;
			if(arr[y][x]=='Y') {
				btk(cur+1,c_len+1,y_count+1,cur);
			} else {
				btk(cur+1,c_len+1,y_count,cur);
			}
			visited[y][x] = false;
			return;
		} else {
			visited[y][x] = true;
			if(arr[y][x]=='Y') {
				btk(cur+1,c_len+1,y_count+1,start);
			} else {
				btk(cur+1,c_len+1,y_count,start);
			}
			visited[y][x] = false;
			return;
		}
	}
	static boolean check(int start) {
		int count1 = 0;
		int y = start/5;
		int x = start%5;
		boolean[][] visit = new boolean[5][5];
		Queue<int[]> queue =new LinkedList<>();
		queue.add(new int[] {y,x});
		while(!queue.isEmpty()) {
			int[] tmp = queue.poll();
			int y1 = tmp[0];
			int x1 = tmp[1];
			if(!visit[y1][x1]) {
				count1+=1;
				visit[y1][x1]=true;
				for(int k=0;k<4;k++) {
					int ny = y1+dy[k];
					int nx = x1+dx[k];
					if(0<=nx && nx<5 && 0<=ny && ny<5 && visited[ny][nx] &&!visit[ny][nx]) {
						queue.add(new int[] {ny,nx});
					}
				}
			}
		}
		if(count1==7) {
			return true;
		} else {
			return false;
		}
	}
}

File: 이준영/백준/Gold/1941. 소문난 칠공주/소문난 칠공주.py
# 일단 7개를 고르고 검사를 연결되어 있는지 검사를 하자
# 7개를 고르는 과정에서 Y가 4개이상이면 칠공주가 될 수 없음
# 나머지를 다 골라도 7이 안되면 리턴
# 마지막으로 True가 되는곳에서 탐색 시작하게 바꿈 (전역변수로 관리)
from collections import deque


def btk(cur, l, s_count):  # 현재위치, 고른 길이, s개수, 시작 y좌표, 시작 x좌표
    global s_y, s_x

    y = cur // 5
    x = cur % 5

    if l - s_count >= 4:
        return

    if l == 7:
        check()
        return

    if 25 - cur + l < 7:
        return

    btk(cur + 1, l, s_count)

    visited[y][x] = True

    s_y = y
    s_x = x

    if arr[y][x] == "S":
        btk(cur + 1, l + 1, s_count + 1)
    else:
        btk(cur + 1, l + 1, s_count)

    visited[y][x] = False


def check():
    global cnt

    st = set([(s_y, s_x)])
    tmp_cnt = 1
    dq = deque([(s_y, s_x)])

    while dq:
        y, x = dq.popleft()
        for k in range(4):
            ny = y + dy[k]
            nx = x + dx[k]
            if 0 <= ny < 5 and 0 <= nx < 5 and visited[ny][nx] and (ny, nx) not in st:
                st.add((ny, nx))
                tmp_cnt += 1
                dq.append((ny, nx))

    if tmp_cnt == 7:
        cnt += 1


arr = [list(input()) for _ in range(5)]
visited = [[False] * 5 for _ in range(5)]
dy = [0, 0, 1, -1]
dx = [1, -1, 0, 0]
s_y = -1
s_x = -1
cnt = 0
btk(0, 0, 0)

print(cnt)
